### ISSUE TITLE
Typescript types for connect and Dispatch<A, S> seem broken with react-redux 7 and redux 4

### ISSUE THREAD

**Issue by davnicwil** [reduxjs/react-redux#1246]
**Do you want to request a _feature_ or report a _bug_?**

Report a bug

**What is the current behavior?**

The following code with `redux` 4.0.1 and `react-redux` 7.0.2

```
import React from 'react'
import { connect } from 'react-redux'
import { Dispatch } from 'redux'

type Props = ReturnType<typeof mapState> & ReturnType<typeof mapDispatch>
const Component = (props: Props) => <div>{props.a} {props.b}</div>
const mapState = (state: any) => ({ a: 'a' })
const mapDispatch = (dispatch: Dispatch<any, any>) => ({ b: 'b' })
const Wrapped = connect(mapState, mapDispatch,)(Component)

const Parent = () => <Wrapped /> // <-- Typescript Errors here
```

Produces this Typescript (v3.4.3) error on`<Wrapped />`

> Property 'b' is missing in type '{}' but required in type 'Readonly<Pick<Props, "b">>'

It seems like Typescript is expecting all props defined in `mapDispatch` to be passed as props to `Wrapped`, which is obviously wrong as they should be injected by `connect`, and anyway are only specified as props on the underlying `Component` (I say 'all props' - there's just one prop here, `b`, but if you add more to `mapDispatch` it expects those too). 

This example is purposefully very simple, but it's not related to not using dispatch anywhere, or the objects not being functions, it still happens when those things are true. It's also not related to using `Dispatch<any, any>` - if you pass actual types for Action and State this same error happens.

Weirdly though, it only happens when you use the `Dispatch` type from redux to type the `dispatch` function in `mapDispatch`. If you use `any`, i.e. `const mapDispatch = (dispatch: any) => ({ b: 'b' })`, the error goes away.

Also, the same behaviour is not true of `mapState` - in the example above you can change the type of state from `any` to whatever you want, and the props defined there are (correctly) never expected to be passed into the `Wrapped` component by TS.

These errors don't happen using the last major version of both `redux` (3.7.2) and `react-redux` (6.0.1) - so unless I'm wrong in my expectations of how these types should work, this seems to be a bug introduced with these latest versions.

An obvious change I can see is in the `Dispatch` type, which now expects 2 generic args instead of 1 going from redux 3 to redux 4. But I'm raising the issue here as I think this is probably a bug with the way the typings for `connect` interact with `Dispatch`.

**Comment by markerikson** [comment]
We don't maintain the TS typings ourselves - those are maintained in DefinitelyTyped by the community.

**Comment by davnicwil** [comment]
Ah, yes of course. Thanks.

If anyone lands here looking for a solution to this, @OliverJAsh already raised the issue over on DefinitelyTyped and has submitted a PR for a fix @ https://github.com/DefinitelyTyped/DefinitelyTyped/pull/34498 - hopefully it'll land soon

**Comment by BBenQi** [comment]
> Ah, yes of course. Thanks.
> 
> If anyone lands here looking for a solution to this, @OliverJAsh already raised the issue over on DefinitelyTyped and has submitted a PR for a fix @ [DefinitelyTyped/DefinitelyTyped#34498](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/34498) - hopefully it'll land soon

I also encountered this problem,Have you solved this problem?
